wallet/types/src/keys.rs

```diff
@@ -13,9 +13,109 @@
 use crypto::key::extended::{ExtendedPrivateKey, ExtendedPublicKey};
+use crypto::key::hdkd::child_number::ChildNumber;
+use crypto::key::hdkd::u31::U31;
 use serialization::{Decode, Encode};
 
+/// The index of the receiving key hierarchy
+const BIP32_RECEIVING_INDEX: ChildNumber = ChildNumber::ZERO;
+/// The index of the change key hierarchy
+const BIP32_CHANGE_INDEX: ChildNumber = ChildNumber::ONE;
+
+/// KeyPurpose errors
+#[derive(thiserror::Error, Debug, Eq, PartialEq)]
+pub enum KeyPurposeError {
+    #[error("Could not convert key index to key purpose: {0}")]
+    KeyPurposeConversion(ChildNumber),
+}
+
+/// The usage purpose of a key i.e. if it is for receiving funds or for change
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Encode, Decode)]
+#[repr(u32)]
+#[allow(clippy::unnecessary_cast)]
+pub enum KeyPurpose {
+    /// This is for addresses created for receiving funds that are given to the user
+    ReceiveFunds = BIP32_RECEIVING_INDEX.get_index(),
+    /// This is for the internal usage of the wallet when creating change output for a transaction
+    Change = BIP32_CHANGE_INDEX.get_index(),
+}
+
+impl KeyPurpose {
+    /// All purposes
+    pub const ALL: [KeyPurpose; 2] = [ReceiveFunds, Change];
+    /// The index for receiving funds
+    const RECEIVE_FUNDS_INDEX: ChildNumber = ChildNumber::from_normal(U31::from_u32_with_msb(0).0);
+    /// The index for change addresses/keys
+    const CHANGE_INDEX: ChildNumber = ChildNumber::from_normal(U31::from_u32_with_msb(1).0);
+
+    pub fn get_deterministic_index(&self) -> ChildNumber {
+        match self {
+            ReceiveFunds => Self::RECEIVE_FUNDS_INDEX,
+            Change => Self::CHANGE_INDEX,
+        }
+    }
+}
+
+impl TryFrom<ChildNumber> for KeyPurpose {
+    type Error = KeyPurposeError;
+
+    fn try_from(num: ChildNumber) -> Result<Self, Self::Error> {
+        match num {
+            BIP32_RECEIVING_INDEX => Ok(ReceiveFunds),
+            BIP32_CHANGE_INDEX => Ok(Change),
+            _ => Err(KeyPurposeError::KeyPurposeConversion(num)),
+        }
+    }
+}
+
+/// Struct that holds information for account addresses
+#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord, Encode, Decode)]
+pub struct KeychainUsageState {
+    /// Last used address index. An address might not be used so the corresponding entry would be
+    /// None, otherwise it would be that last used ChildNumber
+    last_used: Option<ChildNumber>,
+    /// Last issued address to the user. Those addresses can be issued until the
+    /// last used index + lookahead size.
+    last_issued: Option<ChildNumber>,
+}
+
+impl KeychainUsageState {
+    pub fn new(last_used: Option<ChildNumber>, last_issued: Option<ChildNumber>) -> Self {
+        Self {
+            last_used,
+            last_issued,
+        }
+    }
+
+    /// Get the last index used in the blockchain
+    pub fn get_last_used(&self) -> Option<ChildNumber> {
+        self.last_used
+    }
```

Comment:
I'd prefer this to be called just `last_used` without `get_`. More consistent with (most of) rest of the code base where `get_` indicates that some work other than returning a field may be performed (e.g. `Idable::get_id` may calculate hash).